api/agent/agent.go

```diff
@@ -661,70 +647,91 @@ func callToHTTPRequest(ctx context.Context, call *call) *http.Request {
 	return req
 }
 
-func (s *hotSlot) dispatch(ctx context.Context, call *call) chan error {
-	// TODO we can't trust that resp.Write doesn't timeout, even if the http
-	// client should respect the request context (right?) so we still need this (right?)
-	errApp := make(chan error, 1)
-
-	go func() {
-		ctx, span := trace.StartSpan(ctx, "agent_dispatch_httpstream")
-		defer span.End()
+func (s *hotSlot) dispatch(ctx context.Context, call *call) error {
+	ctx, span := trace.StartSpan(ctx, "agent_dispatch_httpstream")
+	defer span.End()
 
-		// TODO it's possible we can get rid of this (after getting rid of logs API) - may need for call id/debug mode still
-		// TODO there's a timeout race for swapping this back if the container doesn't get killed for timing out, and don't you forget it
-		swapBack := s.container.swap(call.stderr, call.stderr, &call.Stats)
-		defer swapBack()
+	// TODO it's possible we can get rid of this (after getting rid of logs API) - may need for call id/debug mode still
+	swapBack := s.container.swap(call.stderr, call.stderr, &call.Stats)
+	defer swapBack()
 
-		req := callToHTTPRequest(ctx, call)
-		resp, err := s.container.udsClient.Do(req)
-		if err != nil {
-			common.Logger(ctx).WithError(err).Error("Got error from UDS socket")
-			errApp <- models.ErrFunctionResponse
-			return
+	resp, err := s.container.udsClient.Do(createUDSRequest(ctx, call))
+	if err != nil {
+		// IMPORTANT: Container contract: If http-uds errors/timeout, container cannot continue
+		s.trySetError(err)
+		// first filter out timeouts
+		if ctx.Err() == context.DeadlineExceeded {
+			return context.DeadlineExceeded
 		}
-		common.Logger(ctx).WithField("resp", resp).Debug("Got resp from UDS socket")
+		return models.ErrFunctionResponse
+	}
+	defer resp.Body.Close()
```

Comment:
leaving note to double check - used to have to do `io.Copy(ioutil.Discard, resp.Body)`  and if we get a 504 from fdk i'm assuming we're going to return that code without the body, i forget if they added resp.Body.Close discarding the body, it was at least discussed... just leaving note to double check

Reply:
I think this is safe as per http.Client Do documentation here:
https://golang.org/pkg/net/http/#Client.Do